server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java

```diff
@@ -1817,7 +1822,147 @@ public void publishVolumeCreationUsageEvent(Volume volume) {
+        if (vmId != null) {
+            // serialize VM operation
+            AsyncJobExecutionContext jobContext = AsyncJobExecutionContext.getCurrentExecutionContext();
+            if (jobContext.isJobDispatchedBy(VmWorkConstants.VM_WORK_JOB_DISPATCHER)) {
+                // avoid re-entrance
+
+                VmWorkJobVO placeHolder = null;
+                placeHolder = createPlaceHolderWork(vmId);
+                try {
+                    Pair<String, String> result = orchestrateCheckAndRepairVolume(volumeId, repair);
+                    return result;
+                } finally {
+                    _workJobDao.expunge(placeHolder.getId());
+                }
+            } else {
+                Outcome<Pair> outcome = checkAndRepairVolumeThroughJobQueue(vmId, volumeId, repair);
+
+                try {
+                    outcome.get();
+                } catch (InterruptedException e) {
+                    throw new RuntimeException("Operation is interrupted", e);
+                } catch (java.util.concurrent.ExecutionException e) {
+                    throw new RuntimeException("Execution exception--", e);
+                }
+
+                Object jobResult = _jobMgr.unmarshallResultObject(outcome.getJob());
+                if (jobResult != null) {
+                    if (jobResult instanceof ConcurrentOperationException) {
+                        throw (ConcurrentOperationException)jobResult;
+                    } else if (jobResult instanceof ResourceAllocationException) {
+                        throw (ResourceAllocationException)jobResult;
+                    } else if (jobResult instanceof Throwable) {
+                        throw new RuntimeException("Unexpected exception", (Throwable)jobResult);
+                    }
+                }
+
+                // retrieve the entity url from job result
+                if (jobResult != null && jobResult instanceof Pair) {
+                    return (Pair<String, String>) jobResult;
+                }
+
+                return null;
+            }
+        } else {
+            CheckAndRepairVolumePayload payload = new CheckAndRepairVolumePayload(repair);
+            VolumeInfo volumeInfo = volFactory.getVolume(volumeId);
+            volumeInfo.addPayload(payload);
+
+            Pair<String, String> result = volService.checkAndRepairVolume(volumeInfo);
+            return result;
+        }
+    }
+
+    protected void validationsForCheckVolumeOperation(VolumeVO volume) {
+        Account caller = CallContext.current().getCallingAccount();
+        _accountMgr.checkAccess(caller, null, true, volume);
+
+        Long volumeId = volume.getId();
+        Long vmId = volume.getInstanceId();
+        if (vmId != null) {
+            UserVmVO vm = _userVmDao.findById(vmId);
+            if (vm == null) {
+                throw new InvalidParameterValueException(String.format("VM not found, please check the VM to which this volume %d is attached", volumeId));
```

Comment:
Why should the volume be attached to a VM?

Reply:
It could be a detached volume also. If in case it is attached to VM we need VM details to process the operation through vmworkjob.

Reply:
basically, if there is a VmId there must be a Vm. (so this block is the inline method `handleVmIdPresentOnVolume()` ;) )